src/System.IO.Compression/src/System/IO/Compression/DeflateStream.cs

```diff
@@ -202,14 +202,26 @@ public override long Position
         public override void Flush()
         {
             EnsureNotDisposed();
+            FlushBuffers();
         }
 
         public override Task FlushAsync(CancellationToken cancellationToken)
         {
             EnsureNotDisposed();
-            return cancellationToken.IsCancellationRequested ?
-                Task.FromCanceled(cancellationToken) :
-                Task.CompletedTask;
+            return cancellationToken.IsCancellationRequested ? Task.FromCanceled(cancellationToken) : FlushAsyncCore(cancellationToken);
```

Comment:
>If it's not common, then it's not worth special-casing.

Since the old Flush was an indirection around a ```stream == null``` check, who knows how that might have been abused. Either way, I think it is better practice to add in the check here rather than going through all the trouble of setting up the async operation just to immediately exit because we're decompressing.

>ReadAsync and WriteAsync both throw if _asyncOperations is greater than 0. Should this as well?

I was wondering about that as well. I added some more tests for the case where we perform multiple async operations concurrently and my conclusion is that yes, FlushAsync should check if _asyncOperations == 0 before continuing.  FlushAsync is similar enough to a call to WriteAsync (but with a different flushcode and treatment of the inputbufferhandle), so it should be treated similarly for locking.